### PR TITLE
Dataset v2.0

### PR DESCRIPTION
**Pull request by aliberts**
## What this does
This PR introduces a new format for `LeRobotDataset`, which is accompanied by a new file structure. As these changes are not backward compatible, we increase `CODEBASE_VERSION` from `v1.6` to `v2.0`.

### What do I need to do?
If you already pushed a dataset using `v1.6` of our codebase, you can use the conversion script `lerobot/common/datasets/v2/convert_dataset_v1_to_v2.py` to convert it to the new format.
You will be asked to enter a prompt describing the task performed in the dataset.

Examples for single-task dataset:
```bash
python lerobot/common/datasets/v2/convert_dataset_v1_to_v2.py \
    --repo-id lerobot/aloha_sim_insertion_human_image \
    --task "Insert the peg into the socket."
```

If you recorded your dataset with one of the manipulator robots currently supported in LeRobot (or your own implementation), you can provide its configuration path to add the motor names and robot type to the dataset info using the `--robot-config` option:

```bash
python lerobot/common/datasets/v2/convert_dataset_v1_to_v2.py \
    --repo-id aliberts/koch_tutorial \
    --task "Pick the Lego block and drop it in the box on the right." \
    --robot-config lerobot/configs/robot/koch.yaml
```

For the more complicated cases of one task per episode or multiple tasks per episodes, please refer to the documentation in that script.

### Motivation
Current implementation of our `LeRobotDataset` suffers from a few shortcomings which make it not easy to use on some aspects. Specifically:
- The structure of the files does not accurately reflect the data structure. Our datasets are structured by episodes, which contrasts with a typical ML scenarios with train/val/test splits (although these concepts can still be relevant here). This makes it hard to easily select a subset of episodes from a dataset since the whole dataset has to be downloaded/loaded. Related: [#440](https://github.com/huggingface/lerobot/issues/440)
- Due to the current [hub's limitations](https://huggingface.co/docs/hub/repositories-recommendations), one can not push a dataset with — at most - more than 10k episodes (less if there are multiple cameras).
- The format is not transparent to the user: in order to get information about the content of a dataset, current options are limited to download the entire dataset and inspect it with a custom script, or try to visualize it using our [visualization tool](https://huggingface.co/spaces/lerobot/visualize_dataset). Related: [#383](https://github.com/huggingface/lerobot/issues/383)
- The default file cache system used by `datasets` and `huggingface_hub` makes it not convenient to create datasets locally (with recording). In order to use the newly created files on disk, these libraries check if those files are present in the cache (which they won't) and if not, will download them even though they may already be on disk.
- Some file format used are too framework specific for this format to be more universal (e.g. .safetensors)
- The dataset viewer on the hub is not compatible with our datasets due to `VideoFrame` not yet being integrated into `datasets`.
- The current implementation lacks support for future features that we may want to add such as:
  - Task-tokens-conditioned training
  - Multirobot policies
  - Depth images (Related: [#435](https://github.com/huggingface/lerobot/issues/435))

### Changes
Some of the biggest change come from the new file structure and their content:

```diff
  .
  ├── data
- │   ├── train-00000-of-0001.parquet
+ │   ├── chunk-000
+ │   │   ├── episode_000000.parquet
+ │   │   ├── episode_000001.parquet
+ │   │   ├── episode_000002.parquet
+ │   │   └── ...
+ │   ├── chunk-001
+ │   │   ├── episode_001000.parquet
+ │   │   ├── episode_001001.parquet
+ │   │   ├── episode_001002.parquet
+ │   │   └── ...
+ │   └── ...
- ├── meta_data
+ ├── meta
- │   ├── episode_data_index.safetensors
+ │   ├── episodes.jsonl
  │   ├── info.json
+ │   ├── stats.json
- │   ├── stats.safetensors
+ │   └── tasks.jsonl
  └── videos
+     ├── chunk-000
+     │   ├── observation.images.laptop
      │   │   ├── episode_000000.mp4
      │   │   ├── episode_000001.mp4
      │   │   ├── episode_000002.mp4
      │   │   └── ...
+     │   ├── observation.images.phone
      │   │   ├── episode_000000.mp4
      │   │   ├── episode_000001.mp4
      │   │   ├── episode_000002.mp4
      │   │   └── ...
+     ├── chunk-001
      └── ...
```
Note that this file-based structure is designed to be as versatile as possible. The parquet files are split by episodes (this was already the case for videos) which allows a much more granular control over which episodes one wants to use and download. The structure of the dataset is entirely described in the info.json file, which can be easily downloaded or viewed directly on the hub before downloading any actual data. The type of files used are very simple and do not need complex tools to be read, it only uses `.parquet`, `.json`, `.jsonl` and `.mp4` files (`.md` for the README).

**Added**
- A `LeRobotDataset` can now be called with an `episodes` argument (e.g. `episodes=[1, 10, 12, 40]`) to select a specific subset of episodes by their episode_index. By doing so, only the files corresponding to these episodes will be downloaded (if they're not already on disk). In that case, the `hf_dataset` attribute will only contain data from these episodes, as well as the `episode_data_index`.
- Dataset metadata logic is now handled by the `LeRobotDatasetMetadata` class. This allows to get info about a dataset before loading the data. For example, you could do this:
```python
# Fetch metadata from the hub
metadata = LeRobotDatasetMetadata("lerobot/pusht")

# Calculate train and val episodes
total_episodes = metadata.total_episodes
episodes = list(range(metadata.total_episodes))
num_train_episodes = math.floor(total_episodes * 90 / 100)
train_episodes = episodes[:num_train_episodes]
val_episodes = episodes[num_train_episodes:]

# Load train an val datasets
train_dataset = LeRobotDataset("lerobot/pusht", episodes=train_episodes)
val_dataset = LeRobotDataset("lerobot/pusht", episodes=val_episodes)
```
- Tasks as natural language prompts are now in every datasets and is needed to create one. Every single task of a dataset is listed in the `tasks.json` mapped to its `task_index` which is what's actually stored in parquet files. Using the api, they can be accessed either with `dataset.tasks` to get that mapping or through `dataset.episode_dict[episode_index]["tasks"]` if you're only interested in a particular episode.
- Various information about the structure of the dataset have been added and is now centralized in the `info.json` (keys, shapes, number of episodes, etc.). It serves as a source of truth for what's inside the dataset.
- `episodes.jsonl` contains per-episode information (episode_index, tasks in natural language and episode lengths). This is accessed through the `episode_dict` attribute in the api.
- `LeRobotDataset.create()` allows to create a new dataset from scratch, either for recording data or for porting an existing dataset to the LeRobotDataset format. To that end, new methods are added:
  - `start_image_writter()`: This instantiates an `ImageWriter` in the `image_writer` attribute to write images asynchonously during data recording. This is automatically called during `LeRobotDataset.create()` if specified in the arguments.
  - `stop_image_writter()`: This is to properly stop and remove the `ImageWriter` from the dataset's attributes. **Importantly:** if the `image_writer` has been set to a multiprocess `ImageWriter`, this needs to be called first if you want to pass this dataset into a parallelized DataLoader as the `ImageWriter` class is not pickleable (required for objects to be transfered between processes). This is not needed when instantiating a dataset with `__init__` as the image_writer then is not created.
  - `add_frame()`: Adds a single timestamp data frame to the `episode_buffer`, which keep data in memory temporarily. Note: this will be merged with the `DataBuffer` from [#445](https://github.com/huggingface/lerobot/pull/445) in a subsequent PR.
  - `add_episode()`: Saves the content of the `episode_buffer` to disk and updates metadata for them to be in sync with the contents of the files. This method expects a `task` argument as a string prompt in natural language describing the task performed in the episode. Videos from that episode can optionally be encoded during this phase but it's not mandatory and can be done later in order to give more flexibility on when to do that.
  - `consolidate()`: This will encode videos that have not yet been encoded, clean up the temporary image files, compute dataset statistics, check timestamps are in sync with the `fps` and perform additional sanity checks in the dataset. It needs to be done before uploading the dataset to the hub with `push_to_hub()`.
  - `clear_episode_buffer()`: This can be used to reset the `episode_buffer` (e.g. to discard data from a current recording).

**Changed**
- The logic for checking timestamps and delta_timestamps sync is taken outside of the `__getitem__()` and is now done during `__init__` or `consolidate`. This has the benefit of both saving computation during the `__getitem__()` as well as knowing immediately if there are sync issues with the timestamps.
- The paths for the parquet and video files are now embedded in the `info.json` to allow flexibility and to easily split chunks of files between directories to avoid the hub's limit of files (10k) per folder.
- We now store every datasets (created or downloaded) in `~/.cache/huggingface/lerobot` by default. Changing `root` or setting the `LEROBOT_HOME` env variable allows to change that location. Every call to the `huggingface_hub` download functions like `snapshot_download` or `hf_hub_download` use the `local_dir` argument to that location so that files are not duplicated in cache and to solve the issue of having to download again files already present on disk.
- Refactored the image writing code from `populate_dataset.py` into an `ImageWriter` class.
- `stats.safetensors` is now `stats.json` (the content remains the same but it's unflattened).
- `episode_data_index.safetensors` is removed but the `episode_data_index` is still in the api to map episode_index to indices.

### Performance
In the nominal case (no `delta_timestamp`), `LeRobotDataset.__getitem__()` is on par with the previous version, sometimes slightly improved but still in the same ballpark generally.

`__getitem__()` call time in seconds (average on 10k iterations):
```
repo_id                                 | v1.6   | v2.0  
--------------------------------------- | ------ | ------
lerobot/aloha_sim_insertion_human_image | 0.0036 | 0.0037
lerobot/aloha_sim_insertion_human       | 0.0029 | 0.0027
lerobot/pusht_image                     | 0.0003 | 0.0003
lerobot/pusht                           | 0.0011 | 0.0009
aliberts/koch_tutorial                  | 0.0111 | 0.0106
lerobot/aloha_mobile_cabinet            | 0.0104 | 0.0101
```

<details>
  <summary>Benchmarking code</summary>

  ```python
  from pathlib import Path
  import time
  import torch
  from lerobot.common.datasets.lerobot_dataset import CODEBASE_VERSION, LeRobotDataset

  repo_ids = [
      "lerobot/aloha_sim_insertion_human_image",
      "lerobot/aloha_sim_insertion_human",
      "lerobot/pusht_image",
      "lerobot/pusht",
      "aliberts/koch_tutorial",
      "lerobot/aloha_mobile_cabinet",
  ]
  num_iterations = 10000
  logfile = Path(f"perf_log_{CODEBASE_VERSION}_{num_iterations}.txt")
  with open(logfile, "a") as file:
      file.write(f"__getitem__() call time in seconds (average on {num_iterations} iterations)\n\n")
      file.write(f"repo_id                                 | {CODEBASE_VERSION}  \n")
      file.write("--------------------------------------- | ------\n")

  for repo_id in repo_ids:
      dataset = LeRobotDataset(repo_id=repo_id)
      durations = []
      for i in range(num_iterations):
          start = time.perf_counter()
          item = dataset[i]
          duration = time.perf_counter() - start
          durations.append(duration)

      avg_duration = torch.Tensor(durations).mean()
      results = f"{repo_id} | {avg_duration:.4f}s"
      print(results)
      with open(logfile, "a") as file:
          file.write(results + "\n")
  ```
</details> 

Using `delta_timestamps`, results are more diverse depending on the dataset but still remain in the same ballpark.
`__getitem__()` call time in seconds (average on 10k iterations), `delta_timestamps=[-1/fps, 0, 1/fps]`:
```
repo_id                                 | v1.6   | v2.0  
--------------------------------------- | ------ | ------
lerobot/aloha_sim_insertion_human_image | 0.0176 | 0.0160
lerobot/aloha_sim_insertion_human       | 0.0073 | 0.0068
lerobot/pusht_image                     | 0.0024 | 0.0032
lerobot/pusht                           | 0.0028 | 0.0043
aliberts/koch_tutorial                  | 0.0200 | 0.0184
lerobot/aloha_mobile_cabinet            | 0.0224 | 0.0181
```

<details>
  <summary>Benchmarking code (delta_timestamps)</summary>

  ```python
  from pathlib import Path
  import time
  import torch
  from lerobot.common.datasets.lerobot_dataset import CODEBASE_VERSION, LeRobotDataset

  repo_ids = [
      "lerobot/aloha_sim_insertion_human_image",
      "lerobot/aloha_sim_insertion_human",
      "lerobot/pusht_image",
      "lerobot/pusht",
      "aliberts/koch_tutorial",
      "lerobot/aloha_mobile_cabinet",
  ]
  num_iterations = 10000
  logfile = Path(f"perf_log_{CODEBASE_VERSION}_{num_iterations}.txt")
  with open(logfile, "a") as file:
      file.write(f"__getitem__() call time in seconds (average on {num_iterations} iterations)\n\n")
      file.write(f"repo_id                                 | {CODEBASE_VERSION}  \n")
      file.write("--------------------------------------- | ------\n")

  for repo_id in repo_ids:
      dataset = LeRobotDataset(repo_id=repo_id)
      fps = dataset.fps
      keys = ["observation.state", *dataset.camera_keys]
      delta_timestamps = {key: [-1/fps, 0, 1/fps] for key in keys}
      dataset = LeRobotDataset(repo_id=repo_id, delta_timestamps=delta_timestamps)
      durations = []
      for i in range(num_iterations):
          start = time.perf_counter()
          item = dataset[i]
          duration = time.perf_counter() - start
          durations.append(duration)

      del dataset
      avg_duration = torch.Tensor(durations).mean()
      results = f"{repo_id} | {avg_duration:.4f}s"
      print(results)
      with open(logfile, "a") as file:
          file.write(results + "\n")
  ```
</details>

### Fixes
- Fix a bug in `load_previous_and_future_frames` which didn't actually raise an error when the requested timestamps from `delta_timestamps` did not correspond to actual timestamps in the dataset.
- Various fixes on the datasets have been made:
  - Some tasks already present in some datasets contained strings which were not part of the task (e.g. `"tf.Tensor(b'Do something', shape=(), dtype=string)"`)
  - Some video files were not properly tracked by git lfs
  - Some datasets present a mismatch between the number of episodes in their parquet and the number of video files. This is being investigated [TODO]
    - `lerobot/aloha_mobile_shrimp`
    - `lerobot/aloha_static_battery`
    - `lerobot/aloha_static_fork_pick_up`
    - `lerobot/aloha_static_thread_velcro`
    - `lerobot/uiuc_d3field`
  - `lerobot/viola` is missing video keys [TODO]

## How it was tested
- Adds `tests/fixtures/` in which fixtures and fixtures factories have been added to simplify writing/adding tests. These factories allow the flexibility to create partially mocked objects *on the fly* to be used in tests, while not relying on other components of the codebase that are not meant to be tested in a particular test (e.g. initializing a dataset using hydra).
- Adds `tests/test_image_writer.py`
- Adds `tests/test_delta_timestamps.py`
- Deactivates a bunch of tests which will need to be redesigned and simplified in further PRs.

## How to checkout & try? (for the reviewer)

Use an existing dataset:
```python
from lerobot.common.datasets.lerobot_dataset import LeRobotDataset

REPO_ID = "lerobot/aloha_sim_insertion_human"  # try with '_image' as well

delta_timestamps = {
    "observation.images.top": [-1, -1/50, 0, 25/50],
    "observation.state": [-1, -1/50, 0, 25/50],
}
dataset = LeRobotDataset(repo_id=REPO_ID, delta_timestamps=delta_timestamps)
```

Try out the new feature to select / download specific episodes:
```python
dataset = LeRobotDataset(repo_id=REPO_ID, episodes=[1, 10, 12, 40])
```

You can also create a new dataset:
```python
from lerobot.common.datasets.lerobot_dataset import LeRobotDataset

REPO_ID = "your_hf_username/test_v2"

new_dataset = LeRobotDataset.create(
    repo_id=REPO_ID,
    fps=30,
    robot=robot,
    image_writer_threads_per_camera=1,
)

# TODO
frame = {
    ...
}
new_dataset.add_frame(frame)
new_dataset.add_episode(task="Do something")
new_dataset.consolidate()
```

### CODE REVIEW

File: lerobot/common/datasets/lerobot_dataset.py [hunk 1]
```diff
@@ -21,94 +21,242 @@
...
+        self.download_data = download_data
+        self.video_backend = video_backend if video_backend is not None else "pyav"
+        self.delta_indices = None
+
+        # Load metadata
+        self.root.mkdir(exist_ok=True, parents=True)
+        self._version = get_hub_safe_version(repo_id, CODEBASE_VERSION)
+        self.info = load_info(repo_id, self._version, self.root)
+        self.stats = load_stats(repo_id, self._version, self.root)
+        self.tasks = load_tasks(repo_id, self._version, self.root)
```
Comment (Cadene) [inline lerobot/common/datasets/lerobot_dataset.py:153]: task indices are string

```
dataset.tasks
{'0': 'Insert the peg into the socket.'}
dataset.tasks[0]
Traceback (most recent call last):
  File "<string>", line 1, in <module>
KeyError: 0
dataset.tasks['0']
'Insert the peg into the socket.'
```

File: lerobot/common/datasets/lerobot_dataset.py [hunk 2]
```diff
@@ -21,94 +21,227 @@
...
+
+        Args:
+            repo_id (str): This is the repo id that will be used to fetch the dataset. Locally, the dataset
+                will be stored under root/repo_id.
+            root (Path | None, optional): Local directory to use for downloading/writing files. You can also
+                set the LEROBOT_HOME environment variable to point to a different location. Defaults to
+                '~/.cache/huggingface/lerobot'.
+            episodes (list[int] | None, optional): If specified, this will only load episodes specified by
+                their episode_index in this list. Defaults to None.
+            split (str, optional): _description_. Defaults to "train".
```
[the remark this answers is not in the document]
Comment (aliberts) [reply lerobot/common/datasets/lerobot_dataset.py]: I've removed it (it wasn't used anymore).
I suggest we just allow to keep a notion of split in the `info.json` as I've done in the conversion script:
```json
"splits": {
    "train": "0:50"
}
```

File: lerobot/common/datasets/lerobot_dataset.py (hunk 1 above)
Comment (aliberts) [reply lerobot/common/datasets/lerobot_dataset.py:162]: Fixed in 835ab5a81b360caecd35db34693b08cc8693b09d

**Comment by apockill** [comment]
Hey folks! Awesome work here. I was wondering what the timeline is for merging Dataset 2.0? 

The reason I ask is I'm about to start working on adding support for [Elephant Robotics MyArm M&C](https://shop.elephantrobotics.com/collections/myarm-mc), and I'm not sure if I should target dataset 1.0 or 2.0 initially.

**Comment by aliberts** [comment]
> Hey folks! Awesome work here. I was wondering what the timeline is for merging Dataset 2.0?

@apockill Thank you for your support! The team took some time off and we're just getting back. Hopefully this will be merged very soon (we mainly need to update some more tests now).

> The reason I ask is I'm about to start working on adding support for [Elephant Robotics MyArm M&C](https://shop.elephantrobotics.com/collections/myarm-mc), and I'm not sure if I should target dataset 1.0 or 2.0 initially.

We will refactor how robot classes are structured soon but for now this PR shouldn't have a big impact on adding support for a new robot. The only thing that's being added to robot classes are the [`motor_features` and `camera_features`](https://github.com/huggingface/lerobot/blob/f3630ad91042e9ee323d8522b8371f4c7ab5f8e6/lerobot/common/robot_devices/robots/manipulator.py#L232-L263) for now (this interface might change in the next PR).

**Comment by Cadene** [comment]
TODO after merging: https://github.com/huggingface/lerobot/pull/485

File: lerobot/common/datasets/v2/batch_convert_dataset_v1_to_v2.py [hunk 3]
```diff
@@ -0,0 +1,880 @@
+#!/usr/bin/env python
```
Comment (Cadene) [inline lerobot/common/datasets/v2/batch_convert_dataset_v1_to_v2.py:1]: What is the difference between `batch_convert_dataset_v1_to_v2.py` and `convert_dataset_v1_to_v2.py`?
Maybe worth adding to the header of `batch_convert_dataset_v1_to_v2.py`

File: lerobot/common/datasets/v2/batch_convert_dataset_v1_to_v2.py [hunk 4]
```diff
@@ -0,0 +1,880 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Note: Since the original Aloha datasets don't use shadow motors, you need to comment those out in
+lerobot/configs/robot/aloha.yaml before running this script.
```
Comment (Cadene) [inline lerobot/common/datasets/v2/batch_convert_dataset_v1_to_v2.py:19]: What about mobile Aloha?

**Review by Cadene — APPROVED** [review]
Beautiful work thanks. Left some comments. Hope it helps :)

**Comment by astroyat** [comment]
I tried training using the new dataset and see some errors in compute_stats.py, should d.stats be changed to d.meta.stats?

File: lerobot/common/datasets/compute_stats.py [hunk 5]
```diff
@@ -200,14 +195,106 @@ def aggregate_stats(ls_datasets) -> dict[str, torch.Tensor]:
...
+#     image_stats = {
+#         "min": np.min(images, axis=axes_to_reduce, keepdims=True),
+#         "max": np.max(images, axis=axes_to_reduce, keepdims=True),
+#         "mean": np.mean(images, axis=axes_to_reduce, keepdims=True),
+#         "std": np.std(images, axis=axes_to_reduce, keepdims=True)
+#     }
+#     for key in image_stats:  # squeeze batch dim
+#         image_stats[key] = np.squeeze(image_stats[key], axis=0)
+
+#     return image_stats
```
[the remark this answers is not in the document]
Comment (aliberts) [reply lerobot/common/datasets/compute_stats.py:300]: https://github.com/huggingface/lerobot/pull/521

File: lerobot/common/datasets/v2/batch_convert_dataset_v1_to_v2.py (hunk 3 above)
Comment (aliberts) [reply lerobot/common/datasets/v2/batch_convert_dataset_v1_to_v2.py:1]: Added a comment at the top
```diff
+ This script is for internal use to convert all datasets under the 'lerobot' hub user account to v2.
```

File: tests/fixtures/dataset_factories.py [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+from unittest.mock import patch
+
+import datasets
+import numpy as np
+import PIL.Image
+import pytest
+import torch
+
+from lerobot.common.datasets.lerobot_dataset import CODEBASE_VERSION, LeRobotDataset, LeRobotDatasetMetadata
+from lerobot.common.datasets.utils import (
```
[the remark this answers is not in the document]
Comment (aliberts) [reply tests/fixtures/dataset_factories.py:12]: This part is going to change in the next iteration when removing hydra, not sure we want to spend the time to refactor it here

File: tests/test_datasets.py [hunk 7]
```diff
@@ -33,18 +33,72 @@
...
 
+
+# TODO(aliberts):
+# - [ ] test various attributes & state from init and create
+# - [ ] test init with episodes and check num_frames
+# - [ ] test add_frame
+# - [ ] test add_episode
+# - [ ] test consolidate
+# - [ ] test push_to_hub
+# - [ ] test smaller methods
```
[the remark this answers is not in the document]
Comment (aliberts) [reply tests/test_datasets.py:98]: I want to take care of this is a dedicated PR for refactoring the tests

File: tests/test_datasets.py [hunk 8]
```diff
@@ -117,6 +171,7 @@ def test_factory(env_name, repo_id, policy_name):
 
 
 # TODO(alexander-soare): If you're hunting for savings on testing time, this takes about 5 seconds.
+@pytest.mark.skip("TODO after v2 migration / removing hydra")
```
[the remark this answers is not in the document]
Comment (aliberts) [reply tests/test_datasets.py:174]: > As of now they are using local artifacts from v1. The alternative would be to make it download datasets from the hub, I suggest we take care of this in a future PR. Wdyt?

File: tests/test_datasets.py [hunk 9]
```diff
@@ -149,6 +204,8 @@ def test_multilerobotdataset_frames():
             assert torch.equal(sub_dataset_item[k], dataset_item[k])
 
 
+# TODO(aliberts, rcadene): Refactor and move this to a tests/test_compute_stats.py
+@pytest.mark.skip("TODO after v2 migration / removing hydra")
```
[the remark this answers is not in the document]
Comment (aliberts) [reply tests/test_datasets.py:208]: > As of now they are using local artifacts from v1. The alternative would be to make it download datasets from the hub, I suggest we take care of this in a future PR. Wdyt?

File: tests/test_datasets.py [hunk 10]
```diff
@@ -368,6 +361,7 @@ def load_and_compare(i):
     # load_and_compare(i - 1)
 
 
+@pytest.mark.skip("TODO after v2 migration / removing hydra")
```
[the remark this answers is not in the document]
Comment (aliberts) [reply tests/test_datasets.py:364]: > As of now they are using local artifacts from v1. The alternative would be to make it download datasets from the hub, I suggest we take care of this in a future PR. Wdyt?

File: tests/test_policies.py [hunk 11]
```diff
@@ -351,6 +352,7 @@ def test_normalize(insert_temporal_dim):
     unnormalize(output_batch)
 
 
+@pytest.mark.skip("TODO after v2 migration / removing hydra")
```
[the remark this answers is not in the document]
Comment (aliberts) [reply tests/test_policies.py:355]: > As of now they are using local artifacts from v1. The alternative would be to make it download datasets from the hub, I suggest we take care of this in a future PR. Wdyt?

File: examples/1_load_lerobot_dataset.py [hunk 12]
```diff
@@ -3,78 +3,120 @@
...
-print(f"frames per second used during data collection: {dataset.fps=}")
-print(f"keys to access images from cameras: {dataset.camera_keys=}\n")
+# LeRobotDataset actually wraps an underlying Hugging Face dataset
+# (see https://huggingface.co/docs/datasets for more information).
+print(dataset.hf_dataset)
 
-# Access frame indexes associated to first episode
+# LeRobot datasets also subclasses PyTorch datasets so you can do everything you know and love from working
+# with the latter, like iterating through the dataset.
+# The __get_item__ iterates over the frames of the dataset. Since our datasets are also structured by
```
[the remark this answers is not in the document]
Comment (aliberts) [reply examples/1_load_lerobot_dataset.py]: Fixed https://github.com/huggingface/lerobot/pull/461/commits/23f6c875b5a9bf8f8306fbf915e3a6727ec6e236

File: lerobot/common/datasets/image_writer.py [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+        elif isinstance(image, PIL.Image.Image):
+            img = image
+        else:
+            raise TypeError(f"Unsupported image type: {type(image)}")
+        img.save(fpath)
+    except Exception as e:
+        print(f"Error writing image {fpath}: {e}")
+
+
+def worker_thread_process(queue: queue.Queue):
```
[the remark this answers is not in the document]
Comment (aliberts) [reply lerobot/common/datasets/image_writer.py]: Fixed https://github.com/huggingface/lerobot/pull/461/commits/23f6c875b5a9bf8f8306fbf915e3a6727ec6e236

File: lerobot/common/datasets/lerobot_dataset.py [hunk 14]
```diff
@@ -15,202 +15,942 @@
...
+        license: str | None = "apache-2.0",
+        push_videos: bool = True,
+        private: bool = False,
+        **card_kwargs,
+    ) -> None:
+        if not self.consolidated:
+            raise RuntimeError(
+                "You are trying to upload to the hub a LeRobotDataset that has not been consolidated yet."
+                "Please call the dataset 'consolidate()' method first."
+            )
```
[the remark this answers is not in the document]
Comment (aliberts) [reply lerobot/common/datasets/lerobot_dataset.py]: Fixed https://github.com/huggingface/lerobot/pull/461/commits/23f6c875b5a9bf8f8306fbf915e3a6727ec6e236

File: lerobot/common/datasets/lerobot_dataset.py [hunk 15]
```diff
@@ -15,202 +15,942 @@
...
+                if img_dir.is_dir():
+                    shutil.rmtree(img_dir)
+
+        # Reset the buffer
+        self.episode_buffer = self._create_episode_buffer()
+
+    def start_image_writer(self, num_processes: int = 0, num_threads: int = 4) -> None:
+        if isinstance(self.image_writer, AsyncImageWriter):
+            logging.warning(
+                "You are starting a new AsyncImageWriter that is replacing an already exising one in the dataset."
```
[the remark this answers is not in the document]
Comment (aliberts) [reply lerobot/common/datasets/lerobot_dataset.py]: Fixed https://github.com/huggingface/lerobot/pull/461/commits/23f6c875b5a9bf8f8306fbf915e3a6727ec6e236

File: lerobot/common/datasets/lerobot_dataset.py [hunk 16]
```diff
@@ -15,202 +15,942 @@
...
+        time for video encoding.
+        """
+        if not episode_data:
+            episode_buffer = self.episode_buffer
+
+        episode_length = episode_buffer.pop("size")
+        episode_index = episode_buffer["episode_index"]
+        if episode_index != self.meta.total_episodes:
+            # TODO(aliberts): Add option to use existing episode_index
+            raise NotImplementedError()
```
[the remark this answers is not in the document]
Comment (aliberts) [reply lerobot/common/datasets/lerobot_dataset.py]: Fixed https://github.com/huggingface/lerobot/pull/461/commits/23f6c875b5a9bf8f8306fbf915e3a6727ec6e236

File: lerobot/common/datasets/utils.py [hunk 17]
```diff
@@ -80,27 +182,69 @@ def hf_transform_to_torch(items_dict: dict[torch.Tensor | None]):
...
+    current_major, _ = _get_major_minor(current_version)
+    major_to_check, _ = _get_major_minor(version_to_check)
+    if major_to_check < current_major and enforce_breaking_major:
+        raise BackwardCompatibilityError(repo_id, version_to_check)
+    elif float(version_to_check.strip("v")) < float(current_version.strip("v")):
+        warnings.warn(
+            f"""The dataset you requested ({repo_id}) was created with a previous version ({version_to_check}) of the
+            codebase. The current codebase version is {current_version}. You should be fine since
+            backward compatibility is maintained. If you encounter a problem, contact LeRobot maintainers on
+            Discord ('https://discord.com/invite/s3KuuzsPFb') or open an issue on github.""",
```
[the remark this answers is not in the document]
Comment (aliberts) [reply lerobot/common/datasets/utils.py:233]: Fixed https://github.com/huggingface/lerobot/pull/461/commits/23f6c875b5a9bf8f8306fbf915e3a6727ec6e236

File: lerobot/common/datasets/v2/convert_dataset_v1_to_v2.py [hunk 18]
```diff
@@ -0,0 +1,666 @@
...
+    parser.add_argument(
+        "--local-dir",
+        type=Path,
+        default=None,
+        help="Local directory to store the dataset during conversion. Defaults to /tmp/lerobot_dataset_v2",
+    )
+    parser.add_argument(
+        "--license",
+        type=str,
+        default="mit",
```
[the remark this answers is not in the document]
Comment (aliberts) [reply lerobot/common/datasets/v2/convert_dataset_v1_to_v2.py]: Fixed https://github.com/huggingface/lerobot/pull/461/commits/23f6c875b5a9bf8f8306fbf915e3a6727ec6e236

File: tests/fixtures/defaults.py [hunk 19]
```diff
@@ -0,0 +1,29 @@
+from lerobot.common.datasets.lerobot_dataset import LEROBOT_HOME
+
+LEROBOT_TEST_DIR = LEROBOT_HOME / "_testing"
+DUMMY_REPO_ID = "dummy/repo"
```
[the remark this answers is not in the document]
Comment (aliberts) [reply tests/fixtures/defaults.py]: Fixed https://github.com/huggingface/lerobot/pull/461/commits/23f6c875b5a9bf8f8306fbf915e3a6727ec6e236

File: lerobot/common/datasets/lerobot_dataset.py [hunk 20]
```diff
@@ -15,202 +15,946 @@
...
+
+        if robot is not None:
+            features = get_features_from_robot(robot, use_videos)
+            robot_type = robot.robot_type
+            if not all(cam.fps == fps for cam in robot.cameras.values()):
+                logging.warning(
+                    f"Some cameras in your {robot.robot_type} robot don't have an fps matching the fps of your dataset."
+                    "In this case, frames from lower fps cameras will be repeated to fill in the blanks."
+                )
+        elif robot_type is None or features is None:
```
Comment (dirkmcpherson) [inline lerobot/common/datasets/lerobot_dataset.py:301]: I think this 'or' should be an 'and' based on the message and use.

Comment (aliberts) [reply lerobot/common/datasets/lerobot_dataset.py:301]: Good catch!

File: .github/workflows/test.yml [hunk 21]
```diff
@@ -103,40 +103,40 @@ jobs:
...
-        run: |
-          poetry install --all-extras
-
-      - name: Test end-to-end
-        run: |
-          make test-end-to-end \
-            && rm -rf outputs
+  # TODO(aliberts, rcadene): redesign after v2 migration / removing hydra
+  # end-to-end:
+  #   name: End-to-end
```
[the remark this answers is not in the document]
Comment (aliberts) [reply .github/workflows/test.yml]: I've added them back for now. They are still downloading datasets from the hub, we will change that in the tests refactor PR.

File: lerobot/common/datasets/v2/batch_convert_dataset_v1_to_v2.py (hunk 4 above)
Comment (aliberts) [reply lerobot/common/datasets/v2/batch_convert_dataset_v1_to_v2.py:21]: This is only for datasets recorded with our own scripts from v1 in which we haven't used features from mobile Aloha, so should be an issue. We should however add them in the future in the example conversion script to port external datasets.

File: lerobot/common/datasets/v2/convert_dataset_v1_to_v2.py [hunk 22]
```diff
@@ -0,0 +1,666 @@
...
+            for arm in robot_cfg["follower_arms"]
+            for motor in robot_cfg["follower_arms"][arm]["motors"]
+        ]
+        action_names = [
+            # f"{arm}_{motor}" for arm in ["left", "right"] for motor in robot_cfg["leader_arms"][arm]["motors"]
+            f"{arm}_{motor}" if len(robot_cfg["leader_arms"]) > 1 else motor
+            for arm in robot_cfg["leader_arms"]
+            for motor in robot_cfg["leader_arms"][arm]["motors"]
+        ]
+    # elif robot_cfg["robot_type"] == "stretch3": TODO
```
[the remark this answers is not in the document]
Comment (aliberts) [reply lerobot/common/datasets/v2/convert_dataset_v1_to_v2.py:169]: The actions of Stretch are currently the state of the gamepad controller which is not ideal. We will change that when refactoring the robot classes. In the meantime if stretch3 datasets have been pushed using v1, we just won't add the motor names in them for v2.

**Comment by Kacchan16** [comment]
Hi @aliberts  @Cadene 
Thanks for your amazing work. I have one doubt, I forked lerobot repo and training some policies, now i want to convert to V1.6 to V2.0, but my episodes are .pth format not in parquet format. I check remaining issues, i didn't find anything. right now while conversion it takes only parquet format.
image
Can you please help me here 
Thanks

**Pull request merged**